src/farcasterd/runtime.rs

```diff
@@ -1448,202 +615,18 @@ impl Runtime {
-                                }
-                            }
-                        });
-                    } else {
-                        self.funding_xmr.insert(swap_id, (address, amount, false));
-                    }
-                }
-            },
-
-            Request::FundingCompleted(blockchain) => {
-                let swapid = get_swap_id(&source)?;
-                if match blockchain {
-                    Blockchain::Bitcoin => {
-                        self.funding_btc.remove(&get_swap_id(&source)?).is_some()
-                    }
-                    Blockchain::Monero => self.funding_xmr.remove(&get_swap_id(&source)?).is_some(),
-                } {
-                    self.stats.incr_funded(&blockchain);
-                    info!(
-                        "{} | Your {} funding completed",
-                        swapid.bright_blue_italic(),
-                        blockchain.bright_green_bold()
-                    );
-                }
-            }
-
-            Request::FundingCanceled(blockchain) => {
-                let swapid = get_swap_id(&source)?;
-                match blockchain {
-                    Blockchain::Bitcoin => {
-                        if self.funding_btc.remove(&get_swap_id(&source)?).is_some() {
-                            self.stats.incr_funding_bitcoin_canceled();
-                            info!(
-                                "{} | Your {} funding was canceled",
-                                swapid.bright_blue_italic(),
-                                blockchain.bright_green_bold()
-                            );
-                        }
-                    }
-                    Blockchain::Monero => {
-                        if self.funding_xmr.remove(&get_swap_id(&source)?).is_some() {
-                            self.stats.incr_funding_monero_canceled();
-                            info!(
-                                "{} | Your {} funding was canceled",
-                                swapid.bright_blue_italic(),
-                                blockchain.bright_green_bold()
-                            );
-                        }
-                    }
-                };
-            }
-
             Request::NeedsFunding(Blockchain::Monero) => {
-                let len = self.funding_xmr.len();
-                let res = self
-                    .funding_xmr
+                let funding_infos: Vec<MoneroFundingInfo> = self
+                    .trade_state_machines
+                    .iter()
+                    .filter_map(|tsm| tsm.needs_funding_monero())
+                    .collect();
+                let len = funding_infos.len();
+                let res = funding_infos
                     .iter()
-                    .filter(|(_, (_, _, autofund))| !*autofund)
                     .enumerate()
-                    .map(|(i, (swap_id, (address, amount, _)))| {
-                        let mut res = format!(
```

Comment:
very gratifying to see the pattern of replacing external state aggregation by state machine state pattern/predicate matching repeating everywhere :)